File: agent-packs/eval-pilot/engine/src/evalpilot/runners/copilot.py
"""Copilot CLI :class:`SUTRunner` implementation.

Ported from the monorepo harness's ``evals/_lib/copilot.py``. Drives the
``copilot`` binary non-interactively inside a workspace and returns a
normalised :class:`~evalpilot.runners.base.RunResult`.

Hardening retained from the original:

* Known fatal Windows exit codes are surfaced via ``result.extra["crash"]``
  so a CLI runtime crash isn't confused with an agent-prompt defect.
* The subprocess is launched in its own process group; on timeout (or
  Ctrl-C) the whole tree is killed via ``psutil`` so xdist workers and the
  CLI's Node subprocess aren't orphaned.
* The prompt is fed via **stdin**, never ``-p``: the Windows ``copilot.CMD``
  shim truncates ``-p`` values at the first newline.
* ``--allow-all --no-ask-user`` is the only flag combo that reliably grants
  write/shell perms in non-interactive mode.
"""

from __future__ import annotations

import os
import shutil
import signal
import subprocess
import sys
import time
from pathlib import Path
from typing import Optional, Sequence

from .base import RunResult, SUTRunner, SUTUnavailable, register_runner


# Known fatal Windows NTSTATUS exit codes (runtime crashes, not agent defects).
_WIN_FATAL_EXIT_CODES: dict[int, str] = {
    3221225477: "STATUS_ACCESS_VIOLATION (0xC0000005)",
    3221225725: "STATUS_STACK_OVERFLOW (0xC00000FD)",
    3221226505: "STATUS_STACK_BUFFER_OVERRUN (0xC0000409)",
}


class CopilotNotInstalled(SUTUnavailable):
    """Raised when no ``copilot`` binary can be found on PATH."""


def _truthy_env(name: str) -> bool:
    return os.environ.get(name, "").strip().lower() in {"1", "true", "yes", "on"}


def _skip_sut() -> bool:
    return _truthy_env("EVALPILOT_SKIP_SUT") or _truthy_env("EVALS_SKIP_SUT")


def _resolve_sut_timeout(requested: float) -> float:
    """Clamp ``requested`` to ``EVALPILOT_SUT_TIMEOUT`` when that env is set."""
    raw = (
        os.environ.get("EVALPILOT_SUT_TIMEOUT")
        or os.environ.get("EVALS_SUT_TIMEOUT")
        or ""
    ).strip()
    if not raw:
        return requested
    try:
        cap = float(raw)
    except ValueError:
        return requested
    return min(requested, cap) if cap > 0 else requested


def find_copilot_bin() -> str:
    """Return absolute path to the ``copilot`` binary or raise."""
    override = os.environ.get("COPILOT_BIN")
    if override:
        return override
    candidate = shutil.which("copilot")
    if not candidate:
        raise CopilotNotInstalled(
            "No `copilot` binary found on PATH. Set COPILOT_BIN to override "
            "(useful for tests with a stubbed binary)."
        )
    return candidate


def _kill_process_tree(pid: int, *, grace_seconds: float = 5.0) -> None:
    """Best-effort: terminate then kill ``pid`` and all descendants."""
    try:
        import psutil  # type: ignore
    except ImportError:
        try:
            os.kill(pid, signal.SIGTERM)
        except (ProcessLookupError, PermissionError, OSError):
            pass
        return

    try:
        parent = psutil.Process(pid)
    except psutil.NoSuchProcess:
        return
    procs = parent.children(recursive=True) + [parent]
    for p in procs:
        try:
            p.terminate()
        except (psutil.NoSuchProcess, psutil.AccessDenied):
            pass
    _, alive = psutil.wait_procs(procs, timeout=grace_seconds)
    for p in alive:
        try:
            p.kill()
        except (psutil.NoSuchProcess, psutil.AccessDenied):
            pass
    psutil.wait_procs(alive, timeout=2.0)


def _popen_kwargs_for_tree_kill() -> dict:
    if sys.platform == "win32":
        return {"creationflags": 0x00000200}  # CREATE_NEW_PROCESS_GROUP
    return {"start_new_session": True}


def _run(
    cmd: Sequence[str],
    *,
    cwd: Path,
    log_path: Path,
    timeout: float,
    stdin_text: Optional[str] = None,
) -> RunResult:
    log_path.parent.mkdir(parents=True, exist_ok=True)
    started = time.monotonic()
    stdout = ""
    stderr = ""
    returncode = 0
    timed_out = False

    proc = subprocess.Popen(
        list(cmd),
        cwd=str(cwd),
        stdin=subprocess.PIPE if stdin_text is not None else subprocess.DEVNULL,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        encoding="utf-8",
        errors="replace",
        **_popen_kwargs_for_tree_kill(),
    )
    try:
        try:
            stdout, stderr = proc.communicate(input=stdin_text, timeout=timeout)
            returncode = proc.returncode
        except subprocess.TimeoutExpired:
            timed_out = True
            _kill_process_tree(proc.pid)
            try:
                stdout, stderr = proc.communicate(timeout=5.0)
            except subprocess.TimeoutExpired:
                proc.kill()
                stdout, stderr = "", ""
            stderr = (stderr or "") + f"\n[evalpilot] TIMEOUT after {timeout}s\n"
            returncode = 124
    except KeyboardInterrupt:
        _kill_process_tree(proc.pid)
        raise

    duration = time.monotonic() - started

    crash = None
    if not timed_out and returncode in _WIN_FATAL_EXIT_CODES:
        crash = _WIN_FATAL_EXIT_CODES[returncode]

    crash_line = f"[crash] {crash}\n" if crash else ""
    log_path.write_text(
        f"$ {' '.join(cmd)}\n[cwd] {cwd}\n[exit] {returncode}\n"
        f"[duration_s] {duration:.2f}\n{crash_line}\n"
        f"--- PROMPT (stdin) ---\n{stdin_text or ''}\n"
        f"--- STDOUT ---\n{stdout}\n--- STDERR ---\n{stderr}\n",
        encoding="utf-8",
        errors="replace",
    )
    return RunResult(
        returncode=returncode,
        stdout=stdout,
        stderr=stderr,
        duration_seconds=duration,
        log_path=log_path,
        timed_out=timed_out,
        extra={"crash": crash} if crash else {},
    )


def _skipped_result(*, cmd: Sequence[str], cwd: Path, log_path: Path,
                    stdin_text: Optional[str]) -> RunResult:
    log_path.parent.mkdir(parents=True, exist_ok=True)
    log_path.write_text(
        f"$ {' '.join(cmd)}\n[cwd] {cwd}\n[exit] 125\n"
        f"[skipped] EVALPILOT_SKIP_SUT is set; SUT not launched.\n\n"
        f"--- PROMPT (stdin) ---\n{stdin_text or ''}\n"
        f"--- STDOUT ---\n\n--- STDERR ---\n",
        encoding="utf-8",
        errors="replace",
    )
    return RunResult(
        returncode=125,
        stdout="",
        stderr="[evalpilot] EVALPILOT_SKIP_SUT set; SUT not launched.\n",
        duration_seconds=0.0,
        log_path=log_path,
        skipped=True,
    )


@register_runner
class CopilotRunner(SUTRunner):
    """Drives the ``copilot`` CLI."""

    name = "copilot"

    def available(self) -> bool:
        try:
            find_copilot_bin()
            return True
        except CopilotNotInstalled:
            return False

    def run_agent(
        self,
        *,
        prompt: str,
        workspace: Path,
        agent: Optional[str],
        log_path: Path,
        timeout: float = 600.0,
        extra_args: Sequence[str] = (),
    ) -> RunResult:
        if _skip_sut():
            agent_part = ["--agent", agent] if agent else []
            return _skipped_result(
                cmd=["<copilot>", *agent_part, "--allow-all", "--no-ask-user"],
                cwd=workspace,
                log_path=log_path,
                stdin_text=prompt,
            )
        bin_path = find_copilot_bin()
        cmd: list[str] = [bin_path]
        if agent:
            cmd += ["--agent", agent]
        cmd += ["--allow-all", "--no-ask-user", *extra_args]
        return _run(
            cmd,
            cwd=workspace,
            log_path=log_path,
            timeout=_resolve_sut_timeout(timeout),
            stdin_text=prompt,
        )

    def run_skill(
        self,
        *,
        skill: str,
        prompt: str,
        workspace: Path,
        log_path: Path,
        timeout: float = 300.0,
        extra_args: Sequence[str] = (),
    ) -> RunResult:
        augmented = (
            f"Use the `{skill}` skill to handle the following request. "
            f"Do not invoke any other skill.\n\n{prompt}"
        )
        return self.run_agent(
            prompt=augmented,
            workspace=workspace,
            agent=None,
            log_path=log_path,
            timeout=timeout,
            extra_args=extra_args,
        )


__all__ = ["CopilotRunner", "CopilotNotInstalled", "find_copilot_bin"]

File: agent-packs/eval-pilot/engine-ts/src/runners/copilot.ts
/**
 * Copilot CLI {@link SUTRunner} implementation.
 *
 * Drives the `copilot` binary non-interactively inside a workspace and returns
 * a normalised {@link RunResult}.
 *
 * Hardening retained from the Python original:
 *  - Known fatal Windows exit codes are surfaced via `result.extra.crash` so a
 *    CLI runtime crash isn't confused with an agent-prompt defect.
 *  - The subprocess is launched in its own process group; on timeout the whole
 *    tree is killed via `tree-kill`.
 *  - The prompt is fed via **stdin**, never `-p`: the Windows `copilot.CMD` shim
 *    truncates `-p` values at the first newline.
 *  - `--allow-all --no-ask-user` is the only flag combo that reliably grants
 *    write/shell perms in non-interactive mode.
 */

import { spawn, type SpawnOptions } from "node:child_process";
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import * as path from "node:path";
import treeKill from "tree-kill";
import {
  captureFromFile,
  telemetryEnabled,
  telemetryEnv,
} from "../telemetry/capture.js";
import { emptyTelemetry } from "../telemetry/model.js";
import {
  makeRunResult,
  registerRunner,
  SUTUnavailable,
  type RunAgentArgs,
  type RunResult,
  type RunSkillArgs,
  type SUTRunner,
} from "./base.js";

/** Known fatal Windows NTSTATUS exit codes (runtime crashes, not defects). */
const WIN_FATAL_EXIT_CODES: Record<number, string> = {
  3221225477: "STATUS_ACCESS_VIOLATION (0xC0000005)",
  3221225725: "STATUS_STACK_OVERFLOW (0xC00000FD)",
  3221226505: "STATUS_STACK_BUFFER_OVERRUN (0xC0000409)",
};

export class CopilotNotInstalled extends SUTUnavailable {}

function truthyEnv(name: string): boolean {
  const v = (process.env[name] ?? "").trim().toLowerCase();
  return v === "1" || v === "true" || v === "yes" || v === "on";
}

function skipSut(): boolean {
  return truthyEnv("EVALPILOT_SKIP_SUT") || truthyEnv("EVALS_SKIP_SUT");
}

/**
 * Resolve the effective SUT wall-clock timeout (seconds).
 *
 * Precedence:
 *   1. Explicit `--sut-timeout` override (`override`, when > 0) — authoritative;
 *      RAISES or lowers the timeout regardless of the frontmatter value.
 *   2. `EVALPILOT_SUT_TIMEOUT` / `EVALS_SUT_TIMEOUT` env — a MAX **cap**; it can
 *      only LOWER `requested`, never raise it (back-compat behaviour retained).
 *   3. `requested` — the spec's frontmatter `timeout:` (or the caller default).
 *
 * When neither override nor env is set, behaviour is unchanged (returns
 * `requested`).
 */
function resolveSutTimeout(requested: number, override?: number | null): number {
  // 1. Explicit CLI override wins and may raise above the frontmatter value.
  if (override != null && !Number.isNaN(override) && override > 0) {
    return override;
  }
  // 2. Env var is a cap only (Math.min): it can lower but never raise.
  const raw = (
    process.env.EVALPILOT_SUT_TIMEOUT ??
    process.env.EVALS_SUT_TIMEOUT ??
    ""
  ).trim();
  if (!raw) return requested;
  const cap = Number(raw);
  if (Number.isNaN(cap)) return requested;
  return cap > 0 ? Math.min(requested, cap) : requested;
}

/** Locate an executable on PATH (honours PATHEXT on Windows). */
function which(cmd: string): string | null {
  const isWin = process.platform === "win32";
  const exts = isWin
    ? (process.env.PATHEXT ?? ".COM;.EXE;.BAT;.CMD").split(";")
    : [""];
  const dirs = (process.env.PATH ?? "").split(path.delimiter);
  for (const dir of dirs) {
    if (!dir) continue;
    for (const ext of exts) {
      const candidate = path.join(dir, cmd + ext);
      if (existsSync(candidate)) return candidate;
    }
  }
  return null;
}

/** Return absolute path to the `copilot` binary or throw. */
export function findCopilotBin(): string {
  const override = process.env.COPILOT_BIN;
  if (override) return override;
  const candidate = which("copilot");
  if (!candidate) {
    throw new CopilotNotInstalled(
      "No `copilot` binary found on PATH. Set COPILOT_BIN to override " +
        "(useful for tests with a stubbed binary).",
    );
  }
  return candidate;
}

function killTree(pid: number): Promise<void> {
  return new Promise((resolve) => {
    treeKill(pid, "SIGTERM", () => resolve());
  });
}

interface RunOpts {
  cwd: string;
  logPath: string;
  timeout: number;
  stdinText?: string | null;
  /** Extra environment variables merged into the child process. */
  env?: Record<string, string> | null;
  /** OTel file-exporter path to parse into telemetry after the run. */
  otelPath?: string | null;
}

function isWindowsBatchWrapper(bin: string): boolean {
  return process.platform === "win32" && /\.(?:cmd|bat)$/i.test(bin);
}

function isWindowsPowerShellScript(bin: string): boolean {
  return process.platform === "win32" && /\.ps1$/i.test(bin);
}

function quoteWindowsShellArg(arg: string): string {
  const escaped = arg
    .replace(/(\\*)"/g, '$1$1\\"')
    .replace(/(\\+)$/g, '$1$1');
  return `"${escaped}"`;
}

function prepareSpawn(
  bin: string,
  args: string[],
  baseOptions: SpawnOptions,
): { bin: string; args: string[]; options: SpawnOptions } {
  if (isWindowsBatchWrapper(bin)) {
    // Node cannot spawn .cmd/.bat directly on Windows; run through cmd.exe
    // and pre-quote each argument so spaces/metacharacters stay in argv.
    return {
      bin,
      args: args.map(quoteWindowsShellArg),
      options: { ...baseOptions, shell: true, detached: false },
    };
  }
  if (isWindowsPowerShellScript(bin)) {
    return {
      bin: "powershell.exe",
      args: ["-NoProfile", "-ExecutionPolicy", "Bypass", "-File", bin, ...args],
      options: { ...baseOptions, detached: false },
    };
  }
  return { bin, args, options: baseOptions };
}

async function runProcess(cmd: string[], opts: RunOpts): Promise<RunResult> {
  mkdirSync(path.dirname(opts.logPath), { recursive: true });
  const started = process.hrtime.bigint();
  const stdinText = opts.stdinText ?? null;

  const [bin, ...args] = cmd;
  const baseOptions: SpawnOptions = {
    cwd: opts.cwd,
    // Own process group so tree-kill can reap the whole tree.
    detached: process.platform !== "win32",
    windowsHide: true,
    env: opts.env ? { ...process.env, ...opts.env } : process.env,
    stdio: [stdinText !== null ? "pipe" : "ignore", "pipe", "pipe"],
  };
  const prepared = prepareSpawn(bin!, args, baseOptions);
  const proc = spawn(prepared.bin, prepared.args, prepared.options);

  let stdout = "";
  let stderr = "";
  let timedOut = false;

  proc.stdout?.setEncoding("utf-8");
  proc.stderr?.setEncoding("utf-8");
  proc.stdout?.on("data", (d) => (stdout += d));
  proc.stderr?.on("data", (d) => (stderr += d));

  if (stdinText !== null && proc.stdin) {
    proc.stdin.write(stdinText);
    proc.stdin.end();
  }

  const returncode: number = await new Promise((resolve) => {
    const timer = setTimeout(() => {
      timedOut = true;
      if (proc.pid) void killTree(proc.pid);
      // Give the tree a moment to die, then resolve with the timeout code.
      setTimeout(() => resolve(124), 5000);
    }, opts.timeout * 1000);

    proc.on("error", () => {
      clearTimeout(timer);
      resolve(-1);
    });
    proc.on("close", (code, signal) => {
      if (timedOut) return; // handled by the timeout path
      clearTimeout(timer);
      if (code === null) {
        // Killed by signal; surface a non-zero code.
        resolve(signal ? 128 : 1);
      } else {
        resolve(code);
      }
    });
  });

  if (timedOut) {
    stderr = (stderr || "") + `\n[evalpilot] TIMEOUT after ${opts.timeout}s\n`;
  }

  const duration = Number(process.hrtime.bigint() - started) / 1e9;

  const crash =
    !timedOut && returncode in WIN_FATAL_EXIT_CODES
      ? WIN_FATAL_EXIT_CODES[returncode]
      : null;

  const crashLine = crash ? `[crash] ${crash}\n` : "";
  writeFileSync(
    opts.logPath,
    `$ ${cmd.join(" ")}\n[cwd] ${opts.cwd}\n[exit] ${returncode}\n` +
      `[duration_s] ${duration.toFixed(2)}\n${crashLine}\n` +
      `--- PROMPT (stdin) ---\n${stdinText ?? ""}\n` +
      `--- STDOUT ---\n${stdout}\n--- STDERR ---\n${stderr}\n`,
    "utf-8",
  );

  return makeRunResult({
    returncode,
    stdout,
    stderr,
    duration_seconds: duration,
    log_path: opts.logPath,
    timed_out: timedOut,
    extra: crash ? { crash } : {},
    telemetry: opts.otelPath
      ? captureFromFile(opts.otelPath)
      : emptyTelemetry("telemetry capture disabled"),
  });
}

function skippedResult(opts: {
  cmd: string[];
  cwd: string;
  logPath: string;
  stdinText?: string | null;
}): RunResult {
  mkdirSync(path.dirname(opts.logPath), { recursive: true });
  writeFileSync(
    opts.logPath,
    `$ ${opts.cmd.join(" ")}\n[cwd] ${opts.cwd}\n[exit] 125\n` +
      `[skipped] EVALPILOT_SKIP_SUT is set; SUT not launched.\n\n` +
      `--- PROMPT (stdin) ---\n${opts.stdinText ?? ""}\n` +
      `--- STDOUT ---\n\n--- STDERR ---\n`,
    "utf-8",
  );
  return makeRunResult({
    returncode: 125,
    stdout: "",
    stderr: "[evalpilot] EVALPILOT_SKIP_SUT set; SUT not launched.\n",
    duration_seconds: 0.0,
    log_path: opts.logPath,
    skipped: true,
  });
}

class CopilotRunner implements SUTRunner {
  readonly name = "copilot";

  available(): boolean {
    try {
      findCopilotBin();
      return true;
    } catch (e) {
      if (e instanceof CopilotNotInstalled) return false;
      throw e;
    }
  }

  async runAgent(args: RunAgentArgs): Promise<RunResult> {
    const extraArgs = args.extra_args ?? [];
    if (skipSut()) {
      const agentPart = args.agent ? ["--agent", args.agent] : [];
      return skippedResult({
        cmd: ["<copilot>", ...agentPart, "--allow-all", "--no-ask-user"],
        cwd: args.workspace,
        logPath: args.log_path,
        stdinText: args.prompt,
      });
    }
    const binPath = findCopilotBin();
    const cmd: string[] = [binPath];
    if (args.agent) cmd.push("--agent", args.agent);
    cmd.push("--allow-all", "--no-ask-user", ...extraArgs);
    const otelPath = telemetryEnabled() ? `${args.log_path}.otel.jsonl` : null;
    return runProcess(cmd, {
      cwd: args.workspace,
      logPath: args.log_path,
      timeout: resolveSutTimeout(args.timeout, args.sut_timeout_override),
      stdinText: args.prompt,
      env: otelPath ? telemetryEnv(otelPath) : null,
      otelPath,
    });
  }

  async runSkill(args: RunSkillArgs): Promise<RunResult> {
    const augmented =
      `Use the \`${args.skill}\` skill to handle the following request. ` +
      `Do not invoke any other skill.\n\n${args.prompt}`;
    return this.runAgent({
      prompt: augmented,
      workspace: args.workspace,
      agent: null,
      log_path: args.log_path,
      timeout: args.timeout,
      extra_args: args.extra_args,
      sut_timeout_override: args.sut_timeout_override,
    });
  }
}

registerRunner("copilot", () => new CopilotRunner());

export { CopilotRunner };

// Exported for unit tests (budget-control regression guards).
export { truthyEnv, skipSut, resolveSutTimeout, runProcess, prepareSpawn };
